File: 백준/Silver/10799. 쇠막대기/쇠막대기.py
bars = input()
stack = []
answer = 0

# 자르기
for i in range(len(bars)):
    # (일때
    if bars[i] == '(':
        stack.append('(')
    # )는 레이저이거나 막대기가 끝남
    else:
        # ()(레이저) -> 자르고 나서 남은 열린괄호(막대기)수 더해주기
        if bars[i-1] == '(':
            stack.pop()
            answer += len(stack)
        # )) -> 막대기가 하나 끝난 것
        else:
            stack.pop() # 하나의 (를 빼준다(막대기 제거)
            answer += 1 # 막대기 토막 하나 더해주기
            
print(answer)

File: 백준/Silver/10799. 쇠막대기/쇠막대기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String ironBar = br.readLine();
        Stack<Integer> stack = new Stack<>();
        int cnt = 0;

        for (int i = 0; i < ironBar.length(); i++) {
            // (일때는 스택에 담는다
            if (ironBar.charAt(i) == '(') {
                stack.push(i);
            } else { // )일때 인덱스 차이가 1이면 레이저 아니면 막대기
                if (i - (stack.peek()) == 1) {
                    stack.pop();
                    cnt += stack.size(); // 잘린 막대기 개수
                } else {
                    stack.pop();
                    cnt += 1; // 잘린 막대기 개수
                }
            }
        }
        System.out.print(cnt);
    }
}